cpp/src/arrow/flight/server.cc

```diff
@@ -892,6 +914,25 @@ Status FlightServerBase::Serve() {
   impl_->old_signal_handlers_.clear();
   impl_->running_instance_ = impl_.get();
 
+#ifndef _WIN32
+  // Create the self-pipe for signal handlers to use
+  if (pipe(impl_->self_pipe_.data()) != 0) {
+    return arrow::internal::IOErrorFromErrno(
+        errno, "Could not initialize self-pipe to wait for signals");
+  }
+  // Make write end nonblocking
+  int flags = fcntl(impl_->self_pipe_[1], F_GETFL);
+  if (flags == -1) {
+    return arrow::internal::IOErrorFromErrno(
+        errno, "Could not initialize self-pipe to wait for signals");
+  }
+  flags |= O_NONBLOCK;
+  if (fcntl(impl_->self_pipe_[1], F_SETFL, flags) == -1) {
+    return arrow::internal::IOErrorFromErrno(
+        errno, "Could not initialize self-pipe to wait for signals");
+  }
+  std::thread handle_signals(&Impl::WaitForSignals, impl_->self_pipe_[0]);
```

Comment:
Ideally this all would be factored in a RAII object (what happens if there is an error below?).